src/uri.c

```diff
@@ -574,3 +557,56 @@ str *coap_get_query(coap_pdu_t *request) {
   }
   return query;
 }
+
+str *coap_get_uri_path(coap_pdu_t *request) {
+  coap_opt_iterator_t opt_iter;
+  coap_opt_filter_t f;
+  coap_opt_t *q;
+  str *uri_path = NULL;
+  size_t length = 0;
+  static const unsigned char hex[] = "0123456789ABCDEF";
+
+  coap_option_filter_clear(f);
+  coap_option_filter_set(f, COAP_OPTION_URI_PATH);
+  coap_option_iterator_init(request, &opt_iter, f);
+  while ((q = coap_option_next(&opt_iter))) {
+    uint16_t seg_len = coap_opt_length(q), i;
+    unsigned char *seg= coap_opt_value(q);
+    for (i = 0; i < seg_len; i++) {
+      if (is_unescaped_in_path(seg[i]))
+	length += 1;
+      else
+	length += 3;
+    }
+    /* bump for the leading "/" */
+    length += 1;
+  }
+  /* The first entry does not have a leading "/" */
+  if (length > 0)
+    length -= 1;
+
+  /* if 0, either no URI_PATH Option, or the first one was empty */
+  uri_path = coap_new_string(length);
+  if (uri_path) {
+    uri_path->length = length;
+    unsigned char *s = uri_path->s;
+    coap_option_iterator_init(request, &opt_iter, f);
+    while ((q = coap_option_next(&opt_iter))) {
+      if (s != uri_path->s)
```

Comment:
This does not work for multiple empty Uri-Path options, as, e.g., in `///`. I propose this instead (tested and works):

```
diff --git a/src/uri.c b/src/uri.c
index 3f30e20..302e6b6 100644
--- a/src/uri.c
+++ b/src/uri.c
@@ -590,10 +590,12 @@ str *coap_get_uri_path(coap_pdu_t *request) {
   if (uri_path) {
     uri_path->length = length;
     unsigned char *s = uri_path->s;
+    int n = 0;
     coap_option_iterator_init(request, &opt_iter, f);
     while ((q = coap_option_next(&opt_iter))) {
-      if (s != uri_path->s)
+      if (n++) {
         *s++ = '/';
+      }
       uint16_t seg_len = coap_opt_length(q), i;
       unsigned char *seg= coap_opt_value(q);
       for (i = 0; i < seg_len; i++) {
```